radix-engine-profiling/src/rocks_db_metrics/tests/write.rs

```diff
@@ -0,0 +1,394 @@
+
+            substate_db.commit(&input_data);
+        }
+
+        // prepare intermediate data
+        for (_k, v) in substate_db.commit_set_metrics.borrow().iter() {
+            for (i, val) in v.iter().enumerate() {
+                let exists = rocksdb_data_intermediate.get(&(i + 1)).is_some();
+                if exists {
+                    rocksdb_data_intermediate
+                        .get_mut(&(i + 1))
+                        .unwrap()
+                        .push(*val);
+                } else {
+                    rocksdb_data_intermediate.insert(i + 1, vec![*val]);
+                }
+            }
+        }
+
+        substate_db.commit_set_metrics.borrow_mut().clear();
+    }
+
+    println!("");
+    // prepare output data
+    let rocksdb_data_output =
+        calculate_percent_to_max_points(&mut rocksdb_data_intermediate, 95f32);
+    // prepare data for plot
+    let mut rocksdb_data = Vec::with_capacity(100000);
+    for (k, v) in &rocksdb_data_intermediate {
+        for val in v {
+            rocksdb_data.push((*k as f32, val.as_micros() as f32));
+        }
+    }
+
+    (rocksdb_data, rocksdb_data_output, rocksdb_data_intermediate)
+}
+
+fn test_commit_per_size_internal<F, S>(
+    rounds_count: usize,
+    min_size: usize,
+    max_size: usize,
+    size_step: usize,
+    prepare_db_write_repeats: usize,
+    create_store: F,
+) -> (
+    Vec<(f32, f32)>,
+    Vec<(f32, f32)>,
+    BTreeMap<usize, Vec<Duration>>,
+)
+where
+    F: Fn(PathBuf) -> SubstateStoreWithMetrics<S>,
+    S: SubstateDatabase + CommittableSubstateDatabase,
+{
+    // RocksDB part
+    let path = PathBuf::from(r"/tmp/radix-scrypto-db");
+    // clean database
+    std::fs::remove_dir_all(path.clone()).ok();
+
+    // prepare database with maxium size
+    {
+        let mut substate_db = create_store(path.clone());
+        prepare_db(
+            &mut substate_db,
+            MIN_SIZE,
+            MAX_SIZE,
+            SIZE_STEP,
+            prepare_db_write_repeats,
+            false,
+        );
+    }
+
+    // reopen database and measure commit times
+    let mut substate_db = create_store(path.clone());
+    let mut rng = rand::thread_rng();
+
+    // prepare vector with substate sizes
+    let mut size_vector: Vec<usize> = Vec::new();
+    for size in (min_size..=max_size).step_by(size_step) {
+        size_vector.push(size);
+    }
+
+    // repeat 1 substate commit n-times
+    for i in 0..rounds_count {
+        print!("Round {}/{}\r", i + 1, rounds_count);
+        std::io::stdout().flush().ok();
+
+        let mut idx_vector = size_vector.clone();
+
+        for _ in 0..size_vector.len() {
+            assert!(!idx_vector.is_empty());
+            // randomize substate size
+            let idx = rng.gen_range(0..idx_vector.len());
+
+            let mut input_data = DatabaseUpdates::new();
+
+            let (partition_key, _sort_key, partition) =
+                generate_commit_data(&mut rng, idx_vector[idx]);
```

Comment:
Again, we're insert into a unique, random partition key, which means we're updating the upper JMT, rather than inserting an entry into an existing lower JMT.

Reply:
Refactored that test, now like for `prepare_db` I am committing various size substates per one node (nodes are generated `rounds_count` times).